chips/nrf52/src/nfct.rs

```diff
@@ -0,0 +1,703 @@
+        rx_amount
+    }
+
+    fn handle_selected(&self) {
+        self.state.set(NfcState::Activated);
+        self.client.map(|client| client.tag_selected());
+    }
+
+    fn handle_rxend(&self) {
+        self.state.set(NfcState::Received);
+        // Return the buffer to the capsule
+        self.client.map(|client| {
+            self.rx_buffer.take().map(|rx_buffer| {
+                let returncode = if self.registers.framestatus_rx.is_set(FrameStatus::OVERRUN) {
+                    ReturnCode::ENOMEM
+                } else if self.registers.framestatus_rx.is_set(FrameStatus::CRCERROR)
+                    || self
+                        .registers
+                        .framestatus_rx
+                        .is_set(FrameStatus::PARITYSTATUS)
+                {
+                    ReturnCode::FAIL
+                } else {
+                    ReturnCode::SUCCESS
+                };
+                let rx_amount = self.get_rx_amount() as usize;
+                self.mask_rx_bits(rx_buffer);
+                client.frame_received(rx_buffer, rx_amount, returncode);
+            });
+        });
+        self.clear_errors();
+    }
+
+    fn handle_txstart(&self) {
+        self.set_tx_registers();
+    }
+
+    fn handle_txend(&self) {
+        self.state.set(NfcState::Transmitted);
+        // Return the buffer to the capsule
+        self.client.map(|client| {
+            self.tx_buffer.take().map(|tx_buffer| {
+                let returncode = if self
+                    .registers
+                    .errorstatus
+                    .is_set(ErrorStatus::FRAMEDELAYTIMEOUT)
+                {
+                    ReturnCode::FAIL
+                } else {
+                    ReturnCode::SUCCESS
+                };
+                client.frame_transmitted(tx_buffer, returncode);
+            });
+        });
+        self.clear_errors();
+    }
+
+    fn clear_errors(&self) {
+        self.registers
+            .errorstatus
+            .set(self.registers.errorstatus.get());
+        self.registers
+            .framestatus_rx
+            .set(self.registers.framestatus_rx.get());
+    }
+
+    pub fn handle_interrupt(&self) {
+        let saved_inter = self.registers.intenset.extract();
+        self.disable_all_interrupts();
+
+        let active_events = self.active_events();
+        let events_to_process = saved_inter.bitand(active_events.get());
+        if events_to_process.is_set(Interrupt::RXFRAMEEND) {
+            self.handle_rxend();
+        }
+        if events_to_process.is_set(Interrupt::TXFRAMEEND) {
+            self.handle_txend();
+        }
+        if events_to_process.is_set(Interrupt::SELECTED) {
+            self.handle_selected();
+        }
+        if events_to_process.is_set(Interrupt::TXFRAMESTART) {
+            self.handle_txstart();
+            // Enable TXFRAMEEND
+            self.registers.intenset.modify(Interrupt::TXFRAMEEND::SET);
+        }
+        self.clear_errors();
```

Comment:
Why do you clear the errors?

Reply:
Those error registers that get cleared are related to transmitting or receiving. They're handled properly in the relevant `TX/RX` functions. And at the end of the `handle_interrupt()` I clear them once more just to ensure there won't be any spurious errors.

I will try to convey that with comments in code.